[UNSPLASHAPP-002] Authentication Screen

IosBlessed opened the pull request:
Created logic + animation for screen if the authorization state of user is unathorized, animated appearance of elements inside block of action section (buttons + privacy text)

IosBlessed committed:
Created logic + animation for screen if the authorization state of user is unathorized, animated appearance of elements inside block of action section (buttons + privacy text)

niclupu commented on unsplahAPI/unsplahAPI/Authentification/Authentication/View/AuthenticationViewController.swift:
```diff
@@ -9,42 +9,161 @@ import UIKit
+            }
+        }
+    }
+    @IBOutlet private weak var createAccountButton: UIButton! {
+        didSet {
+            createAccountButton.setAttributedTitle(
+                NSAttributedString(
+                    string: "Create Account",
+                    attributes: [
+                        NSAttributedString.Key.font: DesignedSystemFonts.bodyBold,
+                        NSAttributedString.Key.foregroundColor: DesignedSystemColors.primaryText
+                    ]),
+                for: .normal
+            )
+        }
+    }
+    @IBOutlet private weak var logInButton: UIButton! {
+        didSet {
+            logInButton.setAttributedTitle(
+                NSAttributedString(
+                    string: "Log In",
+                    attributes: [
+                        NSAttributedString.Key.font: DesignedSystemFonts.bodyBold,
+                        NSAttributedString.Key.foregroundColor: DesignedSystemColors.primaryText
+                    ]),
+                for: .normal
+            )
+        }
+    }
```
Looks like these buttons are the same, let's create a custom button with following attributes and corner radius so we won't have duplicated code

niclupu commented on unsplahAPI/unsplahAPI/Authentification/Authentication/View/AuthenticationViewController.swift:
```diff
@@ -9,42 +9,161 @@ import UIKit
+        self.unitOpacity = 0
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        let transitionOfScrollViewUp: Double = 2.0
+        UIView.animate(withDuration: transitionOfScrollViewUp) { [weak self] in
+            guard let self else { return }
+            self.backgroundImageTopConstraint.constant = -280
+            self.backgroundImageBottomConstraint.constant = 280
+            homeScreenView.layoutIfNeeded()
+        }
+        UIView.animate(withDuration: 1.0, delay: transitionOfScrollViewUp) { [weak self] in
+            guard let self else { return }
+            self.unitOpacity = 1.0
+        }
+    }
```
Let's add to Constants

niclupu commented on unsplahAPI/unsplahAPI/Authentification/Authentication/View/AuthenticationViewController.swift:
```diff
@@ -9,42 +9,161 @@ import UIKit
-       userCancelable = viewModel?.user.$authenticationState.sink { [weak self] autheticationState in
+        userCancelable = viewModel?.user.$authenticationState.sink { [weak self] autheticationState in
             switch autheticationState {
             case .unauthenticated:
-                //TODO: show auth screen
+                // TODO: show auth screen
                 self?.authenticationProcess(result: .failure(AuthenticationErrors.unableToIdentifyUser))
             case .authenticated:
                 self?.authenticationProcess(result: .success(self?.viewModel?.user))
             case .authenticating:
-                //TODO: perform default process
+                // TODO: perform default process
                 print(autheticationState)
```
I'm thinking this logic should be done in ViewModel as ViewController doesn't need to know about autenticationState or user at all, instead we can bind to success and perform navigation when we receive true.

```
class AuthenticationViewModel: AuthenticationViewModelProtocol {    
    
    var error: AnyPublisher<String, Never> {
        return errorSubject.eraseToAnyPublisher()
    }
    private let errorSubject = CurrentValueSubject<String, Never>("")
    
    var shouldNavigateToNextModule: AnyPublisher<Bool, Never> {
        return shouldNavigateToNextModuleSubject.eraseToAnyPublisher()
    }
    private let shouldNavigateToNextModuleSubject = CurrentValueSubject<Bool, Never>(false)
        
    func authenticate(username: String, password: String) {
        if error {
            errorSubject.value = "Failed to register the user"
        } else {
            shouldNavigateToNextModuleSubject.value = true
        }
    }
}
```

IosBlessed replied:
I think, that the viewController should know, what he must present in case if viewModel receives user with the different authentification states

vikabelinschi replied:
This is why we use bindings, the logic from the viewModel will notify the view controller on what should be presented.

IosBlessed replied:
for sure, this is the reason why I published the user's properties, when it changes, viewController will decide, what to present, as I know, it's not dumb class now, in MVVM

niclupu commented on unsplahAPI/unsplahAPI/Authentification/Authentication/View/AuthenticationViewController.swift:
```diff
@@ -9,42 +9,161 @@ import UIKit
+        case .success(let _):
             coordinator?.didFinishAuthentification()
         case .failure(let error):
             self.constructView()
```
I don't think the bellow stackview should depend on authentication, authentication is done when button is tapped.

IosBlessed committed:
Created animated login view with the designed template, realized logic of processing input data inside email and password field, fixed issues which where related to constraints of background image on iPhone with the bigger resolution then 14 pro

IosBlessed committed:
Created screen for creation of account, modified road of interaction via coordinator and modules, fixed bug which was related to background image positioning of Login Screen

IosBlessed committed:
Created logic of forgot password screen, modified logic of interaction insinde MVVM of Forgot Password section, restructured protocol of AuthentificationCoordinator for understanding, where create account section is first responder or second
UPDATES 1.0: Fixed issue that was related to overlaying of the password when it's inputed, refactored algorithm of processing if inputed data is in correct format( realisation moved to the parent class of reusable login's details)
UPDATES 2.0: Restructured whole logic of interaction between modules of MVVM, refused from Combine for the Observable pattern and binding, Restructured algorithm of processing of text fields, when user introduces data. Re-build reusable text fields and fixed error, which where related to design of fields, where been inputted wrong text, linted some extra spaces.

niclupu commented on unsplahAPI/unsplahAPI/Authentification/Authentication/View/AuthenticationViewController.swift:
```diff
@@ -8,45 +8,223 @@
+class AuthenticationViewController: UIViewController, AuthenticationViewControllerProtocol {
+
     var coordinator: AuthentificationCoordinatorProtocol?
     var viewModel: AuthenticationViewModelProtocol?
```
Just a suggestion you so you won't have to work with optionals when we know these properties won't be nil, we could initialise them in vc:
```
    private let coordinator: AuthentificationCoordinatorProtocol
    private let viewModel: AuthenticationViewModelProtocol

    init(viewModel: AuthentificationCoordinatorProtocol, coordinator: AuthenticationViewModelProtocol) {
       self.viewModel = viewModel
       self.coordinator = coordinator
       super.init(nibName: String(describing: BrowseViewController.self), bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

IosBlessed replied:
as I think, if you can make any property as optional, just do it. It's better to make a mistake with the logic than to get an unexpected error code and crash.

niclupu replied:
Optionals should be used when we work with nil values, viewModel will never be nil. This is more related to code quality and readability, why handle optional cases when we know for sure a property won't be nil. Also I don't really see any unexpected behaviour coming with this implementation

niclupu commented on unsplahAPI/unsplahAPI/Authentification/Authentication/View/AuthenticationViewController.swift:
```diff
@@ -8,45 +8,223 @@
     var coordinator: AuthentificationCoordinatorProtocol?
     var viewModel: AuthenticationViewModelProtocol?
-    var userCancelable: AnyCancellable?
+    private var userCancelable: AnyCancellable?
```
Let's keep all the iboutlets on top

niclupu commented on unsplahAPI/unsplahAPI/Authentification/Login/View/LoginViewController.swift:
```diff
@@ -1,29 +1,322 @@
+        loginTextView.passwordTextField.placeholder = "Password"
+        loginTextView.layer.opacity = 0.0
+        return loginTextView
+    }()
```
Let's keep one space between methods

niclupu reviewed (changes requested):
Let's refactor the code a little. 
1. Firstly let's respect the app structure defined in the ticket and there's specified to use Observables. We can use Combine only if it has any advantage, but I don't think it's the case for now.
2. Let's improve MVVM separation of concerns, where viewModel handles the logic, and ViewController updates the UI. Let's keep the following structure and have it consistent across all modules ( not using completion in one module and observables in others, let's have a defined structure of the project).

```
class AuthenticationViewController: UIViewController, AuthenticationViewProtocol {
    var viewModel: AuthenticationViewModelProtocol?
    var coordinator: AuthetificationCoordinatorProtocol?

    override func viewDidLoad() {
        super.viewDidLoad()
        // TODO: Setup view, bindings, etc.
        viewModel?.isAuthenticated.bind { [weak self] isAuthenticated in
            if isAuthenticated {
                self?.authenticationSuccessful()
            }
        }
    }

    func authenticationSuccessful() {
        coordinator?.finishAuthentication()
    }
}

// Authentication View Model
class AuthenticationViewModel: AuthenticationViewModelProtocol {
    var isAuthenticated: Observable<Bool> = Observable(false)

    func authenticate(username: String, password: String) {
        // TODO: Perform authentication
        // Set isAuthenticated to true if authentication succeeds
        // isAuthenticated.value = true
    }
}
```
This code is taken from ticket specifications and the separation of concerns is very explicit there, let's respect it.
